### PR TITLE
Add user authentication and authorization

### PR DESCRIPTION
**Pull request by meo84**
**What?** 
So far my application mixed public and admin pages, without the notion of user access. As I want to deploy the app and get started on continuous integration (\o/), I included a login system to authenticate users and give them access to admin pages if allowed. 
_Note_ There is no signup system as the only users are admin users (the app is only meant to be "viewed" by visitors)

**How?** 
I did the authentication and authorization process manually (secured the password with bcrypt) to understand it before using gems like devise and pundit. 
For the integration tests, I used Capybara with RSpec. 

**Questions?**

- I only setup the integration tests for the authentication process, this way I'll apply any feedback you have on the setup to the tests for the authorization process. In particular, would you use a db seed to create the admin user? 
- At first I was hesitant about creating a User model, since there will only ever be two admin users at most, but couldn't think of another approach. Do you have other approaches in mind that could be more relevant? 
- Any other suggestions welcome :)

### CODE REVIEW

**Review by yaf — COMMENTED** [review]
Don't git add auto-generate coffee files

**Review by yaf — COMMENTED** [review]
Don't add scss auto-generate files

File: app/controllers/directors_controller.rb [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-      		format.json { render json: @directors_podium }
-      		format.html
-    	end
-	end
+  before_action :set_director, only: [:edit, :update]
+  before_action :logged_in_user, only: [:new, :create, :edit, :update]
+  before_action :correct_user, only: [:new, :create, :edit, :update]
+
+  def who_graph
+    @directors_podium = [ Director.rank_by_movie_nb(1), Director.rank_by_movie_nb(2), Director.rank_by_movie_nb(3) ]
```
Comment (yaf) [inline app/controllers/directors_controller.rb]: Perhaps this array could be return by a other file  :-)

File: app/controllers/welcome_controller.rb [hunk 2]
```diff
@@ -1,11 +1,11 @@
 class WelcomeController < ApplicationController
 
-	def home
-		@movies_podium = [ Movie.rank_by_attendance(1), Movie.rank_by_attendance(2), Movie.rank_by_attendance(3)]
-		@oldest_movie = Movie.oldest
-		@latest_movie = Movie.latest
+  def home
+    @movies_podium = [ Movie.rank_by_attendance(1), Movie.rank_by_attendance(2), Movie.rank_by_attendance(3)]
```
Comment (yaf) [inline app/controllers/welcome_controller.rb]: Same as for director, this podium could be move (and test) easily in a method of an other object

File: app/helpers/users_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (yaf) [inline app/helpers/users_helper.rb]: Don't add Generate empty helper

File: spec/helpers/sessions_helper_spec.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (yaf) [inline spec/helpers/sessions_helper_spec.rb]: Remove this generate empty file :-)

File: spec/views/sessions/new.html.erb_spec.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (yaf) [inline spec/views/sessions/new.html.erb_spec.rb]: Remove this generate empty file

File: app/assets/javascripts/users.coffee [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (yaf) [inline app/assets/javascripts/users.coffee]: Remove this empty file

File: app/assets/stylesheets/sessions.scss [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (yaf) [inline app/assets/stylesheets/sessions.scss]: Remove this empty file

File: app/assets/stylesheets/users.scss [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (yaf) [inline app/assets/stylesheets/users.scss]: Remove this empty file

**Review by yaf — CHANGES_REQUESTED** [review]
Some stuff to remove (important part)
And to get a better code, `extract method` from controller to model (or an other object)

File: app/controllers/welcome_controller.rb [hunk 9]
```diff
@@ -1,11 +1,11 @@
 class WelcomeController < ApplicationController
 
-	def home
-		@movies_podium = [ Movie.rank_by_attendance(1), Movie.rank_by_attendance(2), Movie.rank_by_attendance(3)]
-		@oldest_movie = Movie.oldest
-		@latest_movie = Movie.latest
+  def home
+    @movies_podium = Movie.most_attended(3)
```
Comment (yaf) [inline app/controllers/welcome_controller.rb:4]: :+1:

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
+module SessionsHelper
```
Comment (yaf) [inline app/helpers/sessions_helper.rb:1]: :+1:

**Review by yaf — APPROVED** [review]
Il ne manque plus que des tests sur le sessionHelper :-)

File: app/models/movie.rb [hunk 11]
```diff
@@ -19,6 +19,7 @@ def self.rank_by_attendance(rank)
   end
 
   def self.most_attended(number)
+    return [] if all.count == 0
```
Comment (yaf) [inline app/models/movie.rb]: It works, but you probably can do a better job here :-)

File: spec/models/movie_spec.rb [hunk 12]
```diff
@@ -39,9 +39,15 @@
     end
 
     describe ".most_attended" do
+
       it "returns array of three movies with most group participants" do
         expect(Movie.most_attended(3)).to eq [gold_movie, silver_movie, bronze_movie]
       end
+
+      it "returns empty array if there are no movies in the database" do
+        Movie.delete_all
```
Comment (yaf) [inline spec/models/movie_spec.rb]: Interesting caution, but it should not be necessary (otherwise, there is a worry I think).

**Review by yaf — CHANGES_REQUESTED** [review]
To approve, I hope to see better impl of `most_attended` and clean test code (without `delete_all`)

**Review by yaf — APPROVED** [review]
:+1: :shipit:

**Comment by meo84** [comment]
Thank you for the review! 👍

**Pull request merged**